File: src/main/java/com/example/leaderboard/domain/user/service/UserService.java
package com.example.leaderboard.domain.user.service;

import com.example.leaderboard.domain.user.dto.UserCreateRequest;
import com.example.leaderboard.domain.user.dto.UserResponse;
import com.example.leaderboard.domain.user.entity.User;
import com.example.leaderboard.domain.user.repository.UserRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;
import java.util.NoSuchElementException;

@Service
@Transactional(readOnly = true)
@RequiredArgsConstructor
public class UserService {

    private final UserRepository userRepository;

    @Transactional
    public UserResponse createUser(UserCreateRequest request) {
        if (userRepository.existsByUsername(request.getUsername())) {
            throw new IllegalArgumentException("이미 사용 중인 username입니다.");
        }
        if (userRepository.existsByEmail(request.getEmail())) {
            throw new IllegalArgumentException("이미 사용 중인 email입니다.");
        }
        User user = new User(
                request.getUsername(),
                request.getEmail(),
                String.valueOf(request.getPassword().hashCode())
        );
        return UserResponse.from(userRepository.save(user));
    }

    public UserResponse getUser(Long id) {
        User user = userRepository.findById(id)
                .orElseThrow(() -> new NoSuchElementException("유저를 찾을 수 없습니다. id=" + id));
        return UserResponse.from(user);
    }

    public List<UserResponse> getUsers() {
        return userRepository.findAll().stream()
                .map(UserResponse::from)
                .toList();
    }
}

File: src/main/kotlin/com/example/leaderboard/domain/user/service/UserService.kt
package com.example.leaderboard.domain.user.service

import com.example.leaderboard.domain.user.dto.UserCreateRequest
import com.example.leaderboard.domain.user.dto.UserResponse
import com.example.leaderboard.domain.user.entity.User
import com.example.leaderboard.domain.user.repository.UserRepository
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

@Service
@Transactional(readOnly = true)
class UserService(private val userRepository: UserRepository) {

    @Transactional
    fun createUser(request: UserCreateRequest): UserResponse {
        if (userRepository.existsByUsername(request.username)) {
            throw IllegalArgumentException("이미 사용 중인 username입니다.")
        }
        if (userRepository.existsByEmail(request.email)) {
            throw IllegalArgumentException("이미 사용 중인 email입니다.")
        }
        // 실제 운영에서는 BCrypt 등 사용. 포트폴리오 단계에서는 단순 해시 대체
        val user = User(
            username = request.username,
            email = request.email,
            passwordHash = request.password.hashCode().toString(),
        )
        return UserResponse.from(userRepository.save(user))
    }

    fun getUser(id: Long): UserResponse {
        val user = userRepository.findById(id)
            .orElseThrow { NoSuchElementException("유저를 찾을 수 없습니다. id=$id") }
        return UserResponse.from(user)
    }

    fun getUsers(): List<UserResponse> =
        userRepository.findAll().map { UserResponse.from(it) }
}
